From d46e37d399eb4d57e04b15e47b074691da06462a Mon Sep 17 00:00:00 2001
From: Jason Hung <jhung@globalscaletechnologies.com>
Date: Fri, 26 Jun 2026 09:27:44 +0800
Subject: [PATCH 1/2] drm/panel: Add support for Startek kd050fhfid103 panel

Add support for the Startek kd050fhfid103 5-inch mipi panel(1920x1080).
---
 drivers/gpu/drm/panel/Kconfig                 |  12 +
 drivers/gpu/drm/panel/Makefile                |   1 +
 .../drm/panel/panel-startek-kd050fhfid103.c   | 560 ++++++++++++++++++
 3 files changed, 573 insertions(+)
 create mode 100644 drivers/gpu/drm/panel/panel-startek-kd050fhfid103.c

diff --git a/drivers/gpu/drm/panel/Kconfig b/drivers/gpu/drm/panel/Kconfig
index e5eba4b05a99..dd9ae6dd0f2f 100644
--- a/drivers/gpu/drm/panel/Kconfig
+++ b/drivers/gpu/drm/panel/Kconfig
@@ -767,6 +767,18 @@ config DRM_PANEL_STARTEK_KD070FHFID078
 	  the host, and a built-in touch controller. Additionally, it requires an
 	  additional PWM for controlling the backlight brightness.
 
+config DRM_PANEL_STARTEK_KD050FHFID103
+	tristate "STARTEK KD050FHFID103 panel"
+	depends on OF
+	depends on DRM_MIPI_DSI
+	depends on BACKLIGHT_CLASS_DEVICE
+	help
+	  Say Y here if you want to enable support for STARTEK KD050FHFID103 DSI panel
+	  based on FORCELEAD-SC5010E controller. The pannel is a 5-inch TFT LCD display
+	  with a resolution of 1920 x 1080 pixels. It provides a MIPI DSI interface to
+	  the host, and a built-in touch controller. Additionally, it requires an
+	  additional PWM for controlling the backlight brightness.
+
 config DRM_PANEL_TDO_TL070WSH30
 	tristate "TDO TL070WSH30 DSI panel"
 	depends on OF
diff --git a/drivers/gpu/drm/panel/Makefile b/drivers/gpu/drm/panel/Makefile
index 2aa36b8cb3de..b5e5f5fed1d3 100644
--- a/drivers/gpu/drm/panel/Makefile
+++ b/drivers/gpu/drm/panel/Makefile
@@ -77,6 +77,7 @@ obj-$(CONFIG_DRM_PANEL_SONY_TD4353_JDI) += panel-sony-td4353-jdi.o
 obj-$(CONFIG_DRM_PANEL_SONY_TULIP_TRULY_NT35521) += panel-sony-tulip-truly-nt35521.o
 obj-$(CONFIG_DRM_PANEL_STARTEK_KD070FHFID015) += panel-startek-kd070fhfid015.o
 obj-$(CONFIG_DRM_PANEL_STARTEK_KD070FHFID078) += panel-startek-kd070fhfid078.o
+obj-$(CONFIG_DRM_PANEL_STARTEK_KD050FHFID103) += panel-startek-kd050fhfid103.o
 obj-$(CONFIG_DRM_PANEL_TDO_TL070WSH30) += panel-tdo-tl070wsh30.o
 obj-$(CONFIG_DRM_PANEL_TPO_TD028TTEC1) += panel-tpo-td028ttec1.o
 obj-$(CONFIG_DRM_PANEL_TPO_TD043MTEA1) += panel-tpo-td043mtea1.o
diff --git a/drivers/gpu/drm/panel/panel-startek-kd050fhfid103.c b/drivers/gpu/drm/panel/panel-startek-kd050fhfid103.c
new file mode 100644
index 000000000000..db97465dfecf
--- /dev/null
+++ b/drivers/gpu/drm/panel/panel-startek-kd050fhfid103.c
@@ -0,0 +1,560 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2018 MediaTek Inc.
+ * Author: Jitao Shi <jitao.shi@mediatek.com>
+ */
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/regulator/consumer.h>
+
+#include <drm/drm_connector.h>
+#include <drm/drm_crtc.h>
+#include <drm/drm_mipi_dsi.h>
+#include <drm/drm_panel.h>
+
+#include <video/mipi_display.h>
+
+struct panel_desc {
+	const struct drm_display_mode *modes;
+	unsigned int bpc;
+
+	/**
+	 * @width_mm: width of the panel's active display area
+	 * @height_mm: height of the panel's active display area
+	 */
+	struct {
+		unsigned int width_mm;
+		unsigned int height_mm;
+	} size;
+
+	unsigned long mode_flags;
+	enum mipi_dsi_pixel_format format;
+	const struct panel_init_cmd *init_cmds;
+	unsigned int lanes;
+};
+
+struct startek_panel {
+	struct drm_panel base;
+	struct mipi_dsi_device *dsi;
+	struct device *dev;
+	const struct panel_desc *desc;
+
+	struct gpio_desc *reset_gpio;
+	struct gpio_desc *enable_gpio;
+	struct gpio_desc *lcmen_gpio;
+
+	bool prepared_power;
+	bool prepared;
+};
+
+enum dsi_cmd_type {
+	INIT_DCS_CMD,
+	INIT_GEN_CMD,
+	DELAY_CMD,
+};
+
+struct panel_init_cmd {
+	enum dsi_cmd_type type;
+	size_t len;
+	const char *data;
+};
+
+#define _INIT_GEN_CMD(...) { \
+	.type = INIT_GEN_CMD, \
+	.len = sizeof((char[]){__VA_ARGS__}), \
+	.data = (char[]){__VA_ARGS__} }
+
+#define _INIT_DCS_CMD(...) { \
+	.type = INIT_DCS_CMD, \
+	.len = sizeof((char[]){__VA_ARGS__}), \
+	.data = (char[]){__VA_ARGS__} }
+
+#define _INIT_DELAY_CMD(...) { \
+	.type = DELAY_CMD,\
+	.len = sizeof((char[]){__VA_ARGS__}), \
+	.data = (char[]){__VA_ARGS__} }
+
+static const struct panel_init_cmd startek_init_cmd[] = {
+	/* P6 */
+	_INIT_GEN_CMD(0x00, 0x68),
+	_INIT_GEN_CMD(0x0F, 0xAA),
+
+	/* P5 */
+	_INIT_GEN_CMD(0x00, 0x58),
+	_INIT_GEN_CMD(0x02, 0x08),
+	_INIT_GEN_CMD(0x03, 0x00),
+	_INIT_GEN_CMD(0x04, 0xA1),
+	_INIT_GEN_CMD(0x05, 0xE2),
+	_INIT_GEN_CMD(0x06, 0x48),
+	_INIT_GEN_CMD(0x07, 0xBF),
+	_INIT_GEN_CMD(0x08, 0xBF),
+	_INIT_GEN_CMD(0x09, 0xFB),
+	_INIT_GEN_CMD(0x0A, 0xBF),
+	_INIT_GEN_CMD(0x0B, 0x30),
+	_INIT_GEN_CMD(0x0C, 0x1C),
+	_INIT_GEN_CMD(0x0D, 0xF1),
+	_INIT_GEN_CMD(0x0E, 0xB2),
+	_INIT_GEN_CMD(0x0F, 0xFC),
+
+	/* P7 */
+	_INIT_GEN_CMD(0x00, 0x78),
+	_INIT_GEN_CMD(0x02, 0x08),
+	_INIT_GEN_CMD(0x03, 0x00),
+	_INIT_GEN_CMD(0x04, 0xA1),
+	_INIT_GEN_CMD(0x05, 0xE2),
+	_INIT_GEN_CMD(0x06, 0x48),
+	_INIT_GEN_CMD(0x07, 0xBF),
+	_INIT_GEN_CMD(0x08, 0xBF),
+	_INIT_GEN_CMD(0x09, 0xFB),
+	_INIT_GEN_CMD(0x0A, 0xBF),
+	_INIT_GEN_CMD(0x0B, 0x30),
+	_INIT_GEN_CMD(0x0C, 0x1C),
+	_INIT_GEN_CMD(0x0D, 0xF1),
+	_INIT_GEN_CMD(0x0E, 0xB2),
+	_INIT_GEN_CMD(0x0F, 0xFC),
+	_INIT_GEN_CMD(0x1A, 0x4D),
+
+	/* P1 */
+	_INIT_GEN_CMD(0x00, 0x18),
+	_INIT_GEN_CMD(0x01, 0x10),
+	_INIT_GEN_CMD(0x02, 0x02),
+	_INIT_GEN_CMD(0x03, 0x00),
+	_INIT_GEN_CMD(0x04, 0xB0),
+	_INIT_GEN_CMD(0x05, 0xA0),
+	_INIT_GEN_CMD(0x06, 0xB9),
+	_INIT_GEN_CMD(0x07, 0x38),
+	_INIT_GEN_CMD(0x08, 0xE0),
+	_INIT_GEN_CMD(0x09, 0x45),
+	_INIT_GEN_CMD(0x0A, 0x32),
+	_INIT_GEN_CMD(0x0B, 0x20),
+	_INIT_GEN_CMD(0x0C, 0xD0),
+	_INIT_GEN_CMD(0x0D, 0xA0),
+	_INIT_GEN_CMD(0x0E, 0x80),
+	_INIT_GEN_CMD(0x0F, 0x25),
+	_INIT_GEN_CMD(0x10, 0x0C),
+	_INIT_GEN_CMD(0x11, 0x00),
+	_INIT_GEN_CMD(0x12, 0x10),
+	_INIT_GEN_CMD(0x17, 0xB0),
+	_INIT_GEN_CMD(0x18, 0xC3),
+	_INIT_GEN_CMD(0x19, 0x0C),
+	_INIT_GEN_CMD(0x1D, 0x0F),
+	_INIT_GEN_CMD(0x1E, 0x0E),
+	_INIT_GEN_CMD(0x1F, 0x00),
+	_INIT_GEN_CMD(0x20, 0x00),
+	_INIT_GEN_CMD(0x24, 0x11),
+	_INIT_GEN_CMD(0x27, 0x00),
+	_INIT_GEN_CMD(0x28, 0x00),
+	_INIT_GEN_CMD(0x29, 0x10),
+	_INIT_GEN_CMD(0x2A, 0x40),
+	_INIT_GEN_CMD(0x2E, 0x0C),
+
+	/* Continue P1 */
+	_INIT_GEN_CMD(0x00, 0x18),
+	_INIT_GEN_CMD(0x40, 0x30),
+	_INIT_GEN_CMD(0x46, 0xC0),
+	_INIT_GEN_CMD(0x48, 0xC2),
+	_INIT_GEN_CMD(0x49, 0x08),
+	_INIT_GEN_CMD(0x4A, 0x08),
+	_INIT_GEN_CMD(0x4B, 0x20),
+	_INIT_GEN_CMD(0x4C, 0x08),
+	_INIT_GEN_CMD(0x4D, 0x08),
+	_INIT_GEN_CMD(0x4F, 0x44),
+	_INIT_GEN_CMD(0x51, 0x00),
+	_INIT_GEN_CMD(0x52, 0x1F),
+
+	/* ENG */
+	_INIT_GEN_CMD(0x00, 0x68),
+	_INIT_GEN_CMD(0x09, 0x94),
+	_INIT_GEN_CMD(0x0A, 0xA2),
+
+	_INIT_GEN_CMD(0x00, 0x98),
+	_INIT_GEN_CMD(0x01, 0x99),
+
+	_INIT_GEN_CMD(0x00, 0x68),
+	_INIT_GEN_CMD(0x09, 0x00),
+	_INIT_GEN_CMD(0x0A, 0x00),
+
+	/* P0 */
+	_INIT_GEN_CMD(0x00, 0x08),
+	//_INIT_GEN_CMD(0x10, 0x45), // BIST_PAT
+	//_INIT_GEN_CMD(0x01, 0x03), // BIST
+	_INIT_GEN_CMD(0x03, 0x77),
+
+	/* Voltage */
+	_INIT_GEN_CMD(0x00, 0x58),
+	_INIT_GEN_CMD(0x13, 0x0A),
+	_INIT_GEN_CMD(0x14, 0x0A),
+	_INIT_GEN_CMD(0x15, 0x04),
+	_INIT_GEN_CMD(0x16, 0x02),
+
+	/* P7 */
+	_INIT_GEN_CMD(0x00, 0x78),
+	_INIT_GEN_CMD(0x13, 0xFF),
+	_INIT_GEN_CMD(0x15, 0x77),
+	_INIT_GEN_CMD(0x16, 0x11),
+	_INIT_GEN_CMD(0x25, 0x34),
+	_INIT_GEN_CMD(0x26, 0x44),
+
+	/* MIPI */
+	_INIT_GEN_CMD(0x00, 0x88),
+	_INIT_GEN_CMD(0x09, 0xFF),
+	_INIT_GEN_CMD(0x20, 0x03),
+	_INIT_GEN_CMD(0x22, 0x42),
+	_INIT_GEN_CMD(0x27, 0x06),
+	//_INIT_GEN_CMD(0x2B, 0xA1), // MIPI STB
+
+	_INIT_DELAY_CMD(20),
+
+	/* OSC */
+	_INIT_GEN_CMD(0x00, 0x98),
+	_INIT_GEN_CMD(0x06, 0xF1),
+
+	/* Sleep Out */
+	_INIT_DCS_CMD(0x11),
+	_INIT_DELAY_CMD(120),
+
+	/* Display On */
+	_INIT_DCS_CMD(0x29),
+	_INIT_DELAY_CMD(20),
+	{},
+};
+
+static inline struct startek_panel *to_startek_panel(struct drm_panel *panel)
+{
+	return container_of(panel, struct startek_panel, base);
+}
+
+static int startek_panel_init_dcs_cmd(struct startek_panel *startek)
+{
+	struct mipi_dsi_device *dsi = startek->dsi;
+	struct drm_panel *panel = &startek->base;
+	int i, err = 0;
+
+	if (startek->desc->init_cmds) {
+		const struct panel_init_cmd *init_cmds = startek->desc->init_cmds;
+
+		for (i = 0; init_cmds[i].len != 0; i++) {
+			const struct panel_init_cmd *cmd = &init_cmds[i];
+
+			switch (cmd->type) {
+			case DELAY_CMD:
+				msleep(cmd->data[0]);
+				err = 0;
+				break;
+
+			case INIT_DCS_CMD:
+				err = mipi_dsi_dcs_write(dsi, cmd->data[0],
+							 cmd->len <= 1 ? NULL :
+							 &cmd->data[1],
+							 cmd->len - 1);
+				break;
+
+			case INIT_GEN_CMD:
+				err = mipi_dsi_generic_write(dsi,
+							 cmd->len <= 1 ? NULL :
+							 &cmd->data[0],
+							 cmd->len);
+				break;
+
+			default:
+				err = -EINVAL;
+			}
+
+			if (err < 0) {
+				dev_err(panel->dev,
+					"failed to write command %u\n", i);
+				return err;
+			}
+		}
+	}
+	return 0;
+}
+
+static int startek_panel_enter_sleep_mode(struct startek_panel *startek)
+{
+	struct mipi_dsi_device *dsi = startek->dsi;
+	int ret;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	ret = mipi_dsi_dcs_set_display_off(dsi);
+	if (ret < 0)
+		return ret;
+
+	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int startek_panel_unprepare_power(struct drm_panel *panel)
+{
+	struct startek_panel *startek = to_startek_panel(panel);
+
+	if (!startek->prepared_power)
+		return 0;
+
+	msleep(150);
+	gpiod_set_value_cansleep(startek->enable_gpio, 0);
+	usleep_range(15000, 20000);
+	gpiod_set_value_cansleep(startek->reset_gpio, 0);
+	usleep_range(15000, 20000);
+	gpiod_set_value_cansleep(startek->lcmen_gpio, 0);
+	msleep(10);
+
+	startek->prepared_power = false;
+
+	return 0;
+}
+
+static int startek_panel_unprepare(struct drm_panel *panel)
+{
+	struct startek_panel *startek = to_startek_panel(panel);
+	int ret;
+
+	if (!startek->prepared)
+		return 0;
+
+	ret = startek_panel_enter_sleep_mode(startek);
+	if (ret < 0) {
+		dev_err(panel->dev, "failed to set panel off: %d\n",
+			ret);
+		return ret;
+	}
+
+	startek_panel_unprepare_power(panel);
+
+	startek->prepared = false;
+
+	return 0;
+}
+
+static int startek_panel_prepare_power(struct drm_panel *panel)
+{
+	struct startek_panel *startek = to_startek_panel(panel);
+	int ret = 0;
+
+	if (startek->prepared_power)
+		return 0;
+
+	gpiod_set_value_cansleep(startek->lcmen_gpio, 0);
+	gpiod_set_value_cansleep(startek->reset_gpio, 0);
+
+	msleep(50);
+	gpiod_set_value_cansleep(startek->lcmen_gpio, 1);
+	gpiod_set_value_cansleep(startek->enable_gpio, 1);
+
+	msleep(50);
+	gpiod_set_value_cansleep(startek->reset_gpio, 1);
+
+	msleep(50);
+	gpiod_set_value_cansleep(startek->reset_gpio, 0);
+	usleep_range(20000, 25000);
+	gpiod_set_value_cansleep(startek->reset_gpio, 1);
+	msleep(150);
+
+	startek->prepared_power = true;
+
+	return 0;
+}
+
+static int startek_panel_prepare(struct drm_panel *panel)
+{
+	struct startek_panel *startek = to_startek_panel(panel);
+	int ret;
+
+	if (startek->prepared)
+		return 0;
+
+	startek_panel_prepare_power(panel);
+
+	ret = startek_panel_init_dcs_cmd(startek);
+	if (ret < 0) {
+		dev_err(panel->dev, "failed to init panel: %d\n", ret);
+		return ret;
+	}
+
+	startek->prepared = true;
+
+	return 0;
+}
+
+static const struct drm_display_mode startek_kd050fhfid103_default_mode = {
+	.clock = 133333,
+	.hdisplay = 1920,
+	.hsync_start = 1920 + 120,
+	.hsync_end = 1920 + 120 + 40,
+	.htotal = 1920 + 120 + 40 + 80,
+	.vdisplay = 1080,
+	.vsync_start = 1080 + 14,
+	.vsync_end = 1080 + 14 + 4,
+	.vtotal = 1080 + 14 + 4 + 31,
+};
+
+static const struct panel_desc startek_kd050fhfid103_desc = {
+	.modes = &startek_kd050fhfid103_default_mode,
+	.bpc = 8,
+	.size = {
+		.width_mm = 116,
+		.height_mm = 72,
+	},
+	.lanes = 4,
+	.format = MIPI_DSI_FMT_RGB888,
+	.mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE |
+		      MIPI_DSI_MODE_LPM,
+	.init_cmds = startek_init_cmd,
+};
+
+static int startek_panel_get_modes(struct drm_panel *panel,
+				struct drm_connector *connector)
+
+{
+	struct startek_panel *startek = to_startek_panel(panel);
+	const struct drm_display_mode *m = startek->desc->modes;
+	struct drm_display_mode *mode;
+
+	mode = drm_mode_duplicate(connector->dev, m);
+	if (!mode) {
+		dev_err(panel->dev, "failed to add mode %ux%u@%u\n",
+			m->hdisplay, m->vdisplay, drm_mode_vrefresh(m));
+		return -ENOMEM;
+	}
+
+	mode->type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED;
+	drm_mode_set_name(mode);
+	drm_mode_probed_add(connector, mode);
+
+	connector->display_info.width_mm = startek->desc->size.width_mm;
+	connector->display_info.height_mm = startek->desc->size.height_mm;
+	connector->display_info.bpc = startek->desc->bpc;
+
+	return 1;
+}
+
+static const struct drm_panel_funcs startek_panel_funcs = {
+	.unprepare = startek_panel_unprepare,
+	.prepare = startek_panel_prepare,
+	.get_modes = startek_panel_get_modes,
+};
+
+static int startek_panel_add(struct startek_panel *startek)
+{
+	struct device *dev = &startek->dsi->dev;
+	int ret;
+
+	startek->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_LOW);
+	if (IS_ERR(startek->reset_gpio)) {
+		ret = PTR_ERR(startek->reset_gpio);
+		dev_err(dev, "cannot get reset-gpios %d\n", ret);
+		return ret;
+	}
+
+	startek->enable_gpio = devm_gpiod_get(dev, "enable", GPIOD_OUT_LOW);
+	if (IS_ERR(startek->enable_gpio)) {
+		ret = PTR_ERR(startek->enable_gpio);
+		dev_err(dev, "cannot get dcdc-en-gpio %d\n", ret);
+		return ret;
+	}
+
+	startek->lcmen_gpio = devm_gpiod_get(dev, "lcmen", GPIOD_OUT_LOW);
+	if (IS_ERR(startek->lcmen_gpio)) {
+		ret = PTR_ERR(startek->lcmen_gpio);
+		dev_err(dev, "cannot get lcmen-gpio %d\n", ret);
+		return ret;
+	}
+
+	drm_panel_init(&startek->base, dev, &startek_panel_funcs,
+			DRM_MODE_CONNECTOR_DSI);
+
+	startek->base.funcs = &startek_panel_funcs;
+	startek->base.dev = &startek->dsi->dev;
+
+	ret = drm_panel_of_backlight(&startek->base);
+	if (ret)
+		return ret;
+
+	drm_panel_add(&startek->base);
+
+	return 0;
+}
+
+static int startek_panel_probe(struct mipi_dsi_device *dsi)
+{
+	struct startek_panel *startek;
+	struct device *dev = &dsi->dev;
+	int ret;
+	const struct panel_desc *desc;
+
+	startek = devm_kzalloc(&dsi->dev, sizeof(*startek), GFP_KERNEL);
+	if (!startek)
+		return -ENOMEM;
+
+	startek->dev = dev;
+	desc = of_device_get_match_data(&dsi->dev);
+	dsi->lanes = desc->lanes;
+	dsi->format = desc->format;
+	dsi->mode_flags = desc->mode_flags;
+	startek->desc = desc;
+	startek->dsi = dsi;
+	ret = startek_panel_add(startek);
+	if (ret < 0)
+		return ret;
+
+	mipi_dsi_set_drvdata(dsi, startek);
+
+	ret = mipi_dsi_attach(dsi);
+	if (ret)
+		drm_panel_remove(&startek->base);
+
+	return ret;
+}
+
+static void startek_panel_shutdown(struct mipi_dsi_device *dsi)
+{
+	struct startek_panel *startek = mipi_dsi_get_drvdata(dsi);
+
+	drm_panel_disable(&startek->base);
+	drm_panel_unprepare(&startek->base);
+}
+
+static void startek_panel_remove(struct mipi_dsi_device *dsi)
+{
+	struct startek_panel *startek = mipi_dsi_get_drvdata(dsi);
+	int ret;
+
+	startek_panel_shutdown(dsi);
+
+	ret = mipi_dsi_detach(dsi);
+	if (ret < 0)
+		dev_err(&dsi->dev, "failed to detach from DSI host: %d\n", ret);
+
+	if (startek->base.dev)
+		drm_panel_remove(&startek->base);
+}
+
+static const struct of_device_id startek_of_match[] = {
+	{ .compatible = "startek,kd050fhfid103",
+	  .data = &startek_kd050fhfid103_desc
+	},
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, startek_of_match);
+
+static struct mipi_dsi_driver startek_panel_driver = {
+	.driver = {
+		.name = "panel-startek-kd050fhfid103",
+		.of_match_table = startek_of_match,
+	},
+	.probe = startek_panel_probe,
+	.remove = startek_panel_remove,
+	.shutdown = startek_panel_shutdown,
+};
+module_mipi_dsi_driver(startek_panel_driver);
+
+MODULE_AUTHOR("Jitao Shi <jitao.shi@mediatek.com>");
+MODULE_DESCRIPTION("STARTEK kd050fhfid103 1920x1080 video mode panel driver");
+MODULE_LICENSE("GPL");
-- 
2.43.0

